cpp/src/arrow/c/bridge.cc

```diff
@@ -587,7 +587,8 @@ struct ArrayExporter {
     export_.buffers_.resize(n_buffers);
     std::transform(buffers_begin, data->buffers.end(), export_.buffers_.begin(),
                    [](const std::shared_ptr<Buffer>& buffer) -> const void* {
-                     return buffer ? buffer->data() : nullptr;
+                     return buffer ? reinterpret_cast<const void*>(buffer->address())
```

Comment:
Do we want to distinguish between C Data Interface and C Device Data Interface here? We could use `buffer->data()` for the former and `buffer->address()` for the latter - ensuring that non-CPU buffers cannot get exported over the C Data Interface.

Reply:
At the moment, this `ArrayExporter` is not aware of whether the ArrowArray came from a plain C Data array or a C Device array. 

I could add that as a boolean flag to the class so we can distinguish based on that here?